### PR TITLE
perf: derive primary-label line/column instead of a second read_span

### PR DESCRIPTION
**Pull request by Boshen**
## Problem

For a single-label diagnostic, `GraphicalReportHandler::render_report` scans the source **twice** from byte 0 to the span:

1. `render_snippets` → `read_span(span, context_lines, context_lines)` — needed for the snippet text.
2. `render_context` → `read_span(primary_label, 0, 0)` — issued **only** to get the primary label's line/column for the `[name:line:col]` header.

`read_span` is O(offset), so on a large file this roughly doubled render cost — the arithmetic gives it away: `render(cal.com.ts) ≈ 731µs ≈ 2 × read_span(cal.com.ts) 344µs`.

## Fix

The primary label always lies inside the `contents` already read for the snippet (guaranteed by the label-in-context filter). So its line/column can be derived by scanning only the short prefix from the context start to the label, instead of rescanning the whole source. The prefix scan uses `memchr2` (the same primitive `context_info` uses), so it stays cheap even on a long / minified line, and its line-break handling mirrors `context_info` (a `\r\n` pair and a lone `\r`/`\n` each count once), so the header is byte-for-byte identical.

Out-of-bounds labels (offset past the payload) return `None` → `fmt::Error`, exactly matching the `OutOfBounds` the old `read_span(label, 0, 0)` produced — not a clamped, misleading position.

## Impact

CodSpeed (instruction counts, the CI perf gate) reports **+71% render improvement**:

| Benchmark | change |
|-----------|-------:|
| `render[cal.com.ts]` (1.4 MB) | **4.1 ms → 2.4 ms** |
| `render[kitchen-sink.tsx]` (716 KB) | **2.1 ms → 1.3 ms** |
| `render_monochrome[cal.com.ts]` | 4.1 ms → 2.4 ms |
| `render_monochrome[kitchen-sink.tsx]` | 2.1 ms → 1.2 ms |

A local stash/restore wall-time A/B agrees (~46% on the same fixtures). Small-file render and `read_span` itself are unchanged.

## Correctness

A differential test (`line_column_tests::matches_read_span_exhaustively`) checks the derived value against a real `read_span(&(off, 0), 0, 0)` across 100k+ randomized offsets over LF / CRLF / lone-CR and multibyte sources — **full equivalence including the error case** (`None` iff the direct read is `OutOfBounds`), sweeping offsets up to two bytes past EOF. This is the same differential approach used for the earlier `read_span` rewrite, and it caught a CRLF edge case (an offset on the `\n` of a `\r\n`, handled by a normalization step). All 38 graphical snapshot tests (which assert the exact header) stay green.

### CODE REVIEW

File: src/handlers/graphical.rs [hunk 1]
```diff
@@ -1415,6 +1411,59 @@ fn split_label(v: &str, style: Style) -> Vec<String> {
...
+/// `base_line` / column `base_column`.
+///
+/// This exists to avoid a second full `read_span` (which re-scans the source
+/// from byte 0) just to locate a label that already lies inside `data`. It
+/// walks only the `data[..rel]` prefix and mirrors the newline handling in
+/// `source_impls::context_info` — a `\r\n` pair and a lone `\r` or `\n` each
+/// count as a single line break — so the result equals
+/// `read_span(&(base_offset + rel, 0).into(), 0, 0)`'s `line()`/`column()`.
+fn line_column_at(data: &[u8], base_line: usize, base_column: usize, rel: usize) -> (usize, usize) {
+    let mut rel = rel.min(data.len());
```
[the remark this answers is not in the document]
Comment (Boshen) [reply src/handlers/graphical.rs]: Good catch — fixed in 199a5f4. `line_column_at` now returns `None` when `rel > data.len()` (the label lies past the payload) instead of clamping, and `render_context` propagates that as `fmt::Error` — restoring the exact `OutOfBounds` behavior of the old `read_span(label, 0, 0)`. The differential test was extended to assert full equivalence including the error case (`None` iff the direct read is `OutOfBounds`), sweeping offsets up to two bytes past EOF, so the `source.len() + 1` empty-span case is now covered.

**Comment by Boshen** [comment]
Consolidated into #219 (this change is its first two commits) so there is a single PR to review and merge.
